gh-124111: Update tkinter for compatibility with Tcl/Tk 9.0.0

culler opened the pull request:
This PR updates tkinter, ttk, test.test_tkinter and test.test_ttk to make them work and pass tests when built against either TclTk 8.6.15, which was released today, or TclTk 9.0.0 which currently has a release candidate and will be released soon.  The changes to the tests include some re-organization to make it easier to accommodate expected development of Tk 9, particularly with respect to the new TclObject of type "pixels" used in Tk 9 for options whose values represent screen distances.

Please also see the discussion on PR #124112.

Note that, unlike #124112, this PR targets Python 3.14.


<!-- gh-issue-number: gh-124111 -->
* Issue: gh-124111
<!-- /gh-issue-number -->

culler committed:
New branch for PR against Python 3.14.

culler committed:
Ttk tests on linux

culler committed:
Remove trailing whitespace.

culler committed:
Add NEWS file.

culler committed:
Make Windows build work; adjust for new changes to Canvas.

zooba commented on PCbuild/tcltk.props:
```diff
@@ -17,15 +17,17 @@
+    <tclLibName>tcl$(TclMajorVersion)$(TclMinorVersion)$(TclDebugExt).lib</tclLibName>
+    <tclShExeName>tclsh$(TclMajorVersion)$(TclMinorVersion)$(TclDebugExt).exe</tclShExeName>
+    <tkDLLName>tcl9tk$(TkMajorVersion)$(TkMinorVersion)$(TclDebugExt).dll</tkDLLName>
+    <tkLibName>tcl9tk$(TkMajorVersion)$(TkMinorVersion)$(TclDebugExt).lib</tkLibName>
```
Are we definitely dropping 8.6 support with this PR? Or do these need to be conditional?

culler replied:
We are definitely not dropping 8.6 support with this PR.  It is meant to support both 8.6 and 9.0.  So, yes, they need to be conditional.

zooba commented on PCbuild/tcltk.props:
```diff
@@ -2,7 +2,7 @@
   <Import Project="pyproject.props" Condition="$(__PyProject_Props_Imported) != 'true'" />
   <PropertyGroup>
-    <TclVersion Condition="$(TclVersion) == ''">8.6.14.0</TclVersion>
+    <TclVersion Condition="$(TclVersion) == ''">9.0.0.0</TclVersion>
```
This should match the version in `get_externals.bat`. It can be overridden with the `TclVersion` environment variable if you're doing a custom build.

culler replied:
I don't know what is the best choice for how to control whether _tkinter is built with Tk 8.6.15 or Tk 9.0.0. 

The way that I created externals/tcltk-9.0.0.0 was to build Tcl 9.0 and Tk 9.0 using nmake in the win directories of those two distributions.  The commands needed were:
`nmake -f makefile.vc`
`nmake -f makefile.vc INSTALLDIR=C\path\to\externals\tcltk-9.0.0.0\amd64 install`

I realize that to make this work in its final form tcltk-9.0.0.0 should be provided by get_externals.bat and there should be some sort of switch that controls which version gets used by build.bat.  Finalizing that design is out of the scope of this PR and beyond my expertise.  This PR is meant to be about changes to _tkinter.c and the test scripts which will be needed to make them work with both TclTk 8.6.15 and 9.0.0.

zooba replied:
Overriding `TclVersion` and `tcltkDir` is the way, but changing the default `TclVersion` as you've done in the PR will break everyone else until we switch fully to relying on 9.0. So I suggest keeping it at 8.6 for now, override it yourself for testing, and when we're ready to switch to 9.0 by default then we can update it (there are more things that have to change before that's possible, so I wouldn't expect it to be in this PR).

zooba reviewed (commented):
Do you know why your generated files look like they're completely changed? It is line endings being bad? (I believe our regen scripts assume you're using autocrlf on Windows.)

culler committed:
Attempt to restore autogenerated files that git tricked me into committing.

zware commented on Lib/test/test_tkinter/test_misc.py:
```diff
@@ -66,9 +71,9 @@ def test_tk_busy(self):
+        with self.assertRaisesRegex(TclError, messages['no_busy']):
             f.tk_busy_configure()
-        with self.assertRaisesRegex(TclError, "can't find busy window"):
+        with self.assertRaisesRegex(TclError, messages['no_busy']):
```
Since we're already using `assertRaisesRegex` here, I think we'd be better off just adjusting the regex rather than maintaining two separate sets of messages:
```suggestion
        with self.assertRaisesRegex(TclError, r"can('| ?no)t find busy window"):
```

Similar elsewhere.

culler replied:
Leaving the PCbuild changes for another PR is fine with me.  I needed to test on Windows.  So I needed to build
_tkinter on Windows.  I was trying to leave some breadcrumbs so others would know what is involved in building against Tk 9 on Windows.  I guess I didn't choose the best way to do that.

zware replied:
Those are valuable breadcrumbs that we don't want to lose :).  I do want us to update the Windows build to 9.0 as soon as we reasonably can, but we'll need to decide whether we're jumping there forever or maintaining support for 8.6 in `PCbuild/` that we ourselves won't be using.  (Personally, I think I'm fine with the direct jump in 3.14, but this is mostly going to be up to @zooba's decision.)

culler replied:
Here is the issue with the regexes, as far as I am concerned.  Unfortunately, there have been many minor changes to the Tk error messages made recently, along the lines of can't -> can not.  That seems to be ongoing.  Maybe it will be easy to write conditional regexes which accept all of the variants.  I wasn't sure.  So I thought that having a catalog might be simpler than relying on regex cleverness.  I agree that, so far, regexes are easily able to accommodate the minor changes to error messages.  I don't know how long the error message tinkering will persist for.

A related topic, which I ignored for this PR, is that the Tcl/Tk project made a big effort to remove the "master-slave" terminology from Tcl and Tk, with considerable internal controversy.  However, tkinter is still using master and slave.  It could be using parent and child instead and be more consistent with Tcl/Tk.

zooba replied:
My current preference is to keep supporting 8.6 for a while, I'm not sure how long for. But since 9.0 isn't officially released yet we certainly can't stop supporting 8.6 yet, and even after they do then we'll want to make sure it's stable and compatible enough for all people who may be building from source (which includes myself, and while we haven't thoroughly audited the 8.6 sources, arguably we should have - being forced into upgrading to 9.0 and re-auditing the sources is a cost that we'd definitely want to defer).

culler replied:
Changes to PCbuild have been reverted.  I renamed the modified versions and added them to the repo, as examples.

zware commented on Lib/test/test_tkinter/test_widgets.py:
```diff
@@ -20,9 +26,14 @@
+    _clipped = {
+        'highlightthickness'} if tk_version < (9, 0) else {
+        'borderwidth', 'height', 'highlightthickness', 'padx', 'pady',
+        'width'}
```
This is pretty much a styling nit, but this might be better done with a base set that is extended if tk is too new:

```suggestion
    _no_round = {'padx', 'pady'}
    _clipped = {'highlightthickness'}
    if tk_version >= (9, 0):
        _no_round.update(
            {
                'borderwidth',
                'height',
                'highlightthickness',
                'width',
            }
        )
        _clipped.update(_no_round)
```

Or as a single

```
if tk_version < (9, 0):
    # set stuff
else:
    # set stuff
```

branch.  The two ternaries back-to-back aren't especially readable to me :)

culler replied:
I wish I knew whether I could count on _no_round being a subset of _clipped.  I think the cases might be more robust at the expense of being more verbose.

culler replied:
Addressed in commit 5a954ecffc.

zware commented on Lib/test/test_tkinter/widget_tests.py:
```diff
@@ -173,7 +180,7 @@ def checkImageParam(self, widget, name):
         self.checkParam(widget, name, image, conv=str)
         self.checkInvalidParam(widget, name, 'spam',
-                errmsg='image "spam" doesn\'t exist')
+                errmsg=messages['no_image'].format('spam'))
```
Similar to above, `checkInvalidParam` uses `assertRaisesRegex` internally:

```suggestion
                errmsg=r"""image "spam" does( no|n')t exist""")
```

culler replied:
I removed the message catalog and replaced it with regexes, where appropriate, or cases based on tk_version.  There
weren't that many places where it was used (yet! - fingers crossed that messing with messages is over).

zware commented on Lib/test/test_ttk/test_widgets.py:
```diff
@@ -479,11 +490,12 @@ def _show_drop_down_listbox(self):
 
     def test_virtual_event(self):
+        if tk_version >= (9, 0) and sys.platform == 'darwin':
+            self.skipTest('test hangs due to Tk 9.0 bug 8b49e9cfa6')
```
This is unfortunate :(

Would you mind to leave a direct link to the bug in a comment here?  And do you have any expectation of when/if it may be fixed, so that we could put an upper bound on the skipped versions?  It doesn't have to be an accurate prediction; we can update the upper bound if it's missed.  I just don't want us to forget about these skips forever :)

culler replied:
I can add a link.  But it turns out not to be a bug, exactly.  There is a change in Tk 9, which is that the combobox menu became a native macOS menu.  It looks much better and it works fine in actual human interaction with the combobox.  But the act of posting a macOS menu causes Apple to create a special event loop which does not pass events to the Tk client until the user dismisses the menu with a mouse click, which may or may not select a menu item.  While this is not an issue with using a combobox, it does create an issue with trying to use generated Tk events to open and close the combobox, as that test does.  The generated events are received and handled by Tk, but Tk currently has no way to dismiss a native macOS menu.  So it has to wait until Apple decides to dismiss the menu based on user interaction.

The choice of not having a "menu unpost" command in the macOS and Window ports of Tk was made intentionally, because those OS's provide their own (user interaction based) ways of dismissing a menu.  X11 provides an unpost command, so Tk wraps that.  In fact, macOS could support a "menu unpost" command.  But that would require a TIP since it would be a change in the Tk API.  If there were such a command then it could be bound to a Tk event and that would allow posting and dismissing combobox menus with event generate commands.

As usual, things that seem simple on the surface can be complex.

Anyway, I think I am going to close the Tk ticket as a "won't fix right now" and consider whether to write a TIP. to add a menu unpost command for Aqua.  I think it just means that that one test will have to be skipped for Tk 9.  I don't see any immediate workaround.

culler replied:
Addressed in commit 5a954ecffc,

culler committed:
Revert other autogenerated files committed by mistake

zware reviewed (commented):
Looking good!  I have some comments that are mostly style nits, and I think we may want to hold off on the changes to `PCbuild/`.  I don't think we'll be able to move the Windows builds for <=3.13 to Tcl/Tk 9 (at least not immediately), so it will make backporting the other changes easier to leave the 3.14 `PCbuild/` updates to a separate PR.

Thanks!

culler committed:
Revert one more autogenerated file committed by mistake

culler committed:
Reference for combobox issue; address conditional definitions of _clipped and _no_round.

culler committed:
Revert PCbuild changes; add examples for tk 9

culler committed:
Adjust no_round for Canvas and Scrollbar.

culler committed:
Adjust _clipped for Canvas once more.

culler committed:
Hmmm = Canvas does clip highlightthickness.

culler committed:
Remove message catalog.

culler committed:
Force checks to rerun since CLA signing seems hung.  (No changes.)

zware commented on Lib/test/test_tkinter/test_widgets.py:
```diff
@@ -7,9 +7,12 @@
+    add_configure_tests,
+    AbstractWidgetTest,
+    StandardOptionsTests,
+    IntegerSizeTests, PixelSizeTests)
```
Nit: let's not stop here :)

```suggestion
    IntegerSizeTests,
    PixelSizeTests,
)
```

culler replied:
OK

zware commented on Lib/test/test_ttk/test_widgets.py:
```diff
@@ -952,25 +975,26 @@ def test_configure_value(self):
 
-
-@add_standard_options(PixelSizeTests if tk_version >= (8, 7) else IntegerSizeTests,
+@add_configure_tests(PixelSizeTests if tk_version >= (8, 7) else IntegerSizeTests,
```
Should be switched to check against `(9, 0)` rather than `(8, 7)`?

culler replied:
I left quite a few tests against (8, 7) in place.  There are many more than just this.  They are not wrong.  They are just useless, assuming that no one ever tries to support using 8.7 with tkinter.  Our understanding is that 8.7 will never be released, so that  assumption is presumably valid.  I guess I could go through all files and replace (8,7) by (9,0) if that seems worth doing.

zware commented on Lib/test/test_tkinter/widget_tests.py:
```diff
@@ -312,8 +356,11 @@ def test_configure_font(self):
+            if tk_version < (9,0):
+                errmsg = 'font "" doesn\'t exist'
+            else:
+                errmsg = 'font "" does not exist'
```
Nit: `checkInvalidParam` uses `assertRaisesRegex`, so we should be able to simplify:
```suggestion
            errmsg = 'font "" does ?n[o\']t exist'
```

culler replied:
OK

culler committed:
Merge branch 'tk9_for_python314' of github.com:culler/cpython into tk9_for_python314

zware commented on Lib/test/test_ttk/test_widgets.py:
```diff
@@ -1147,7 +1185,12 @@ def test_traversal(self):
         self.nb.select(0)
 
-        focus_identify_as = 'focus' if sys.platform != 'darwin' else ''
+        if sys.platform == 'darwin':
+            focus_identify_as = ''
+        elif sys.platform == 'win32':
+            focus_identify_as = 'focus'
+        else:
+            focus_identify_as = 'focus' if tk_version < (9,0) else 'padding'
         self.assertEqual(self.nb.identify(5, 5), focus_identify_as)
```
Trying this locally on Linux with Tcl/Tk 9.0b3, I get a failure where 'focus' is gotten instead of 'padding'.  Maybe we should just adjust the test like the following?

```suggestion
        self.assertIn(self.nb.identify(5, 5), ['', 'focus', 'padding'])
```

culler replied:
I think this is due to a recent change in the size of the focus ring.

culler committed:
Update build files to use Tcl/Tk 8.6.14 by default but to allow any version by setting the TclVersion environment variable

zware reviewed (commented):
Some of these changes cause failures for me on Pop_OS! 22.04 (Ubuntu 22.04 derivative) with Tcl/Tk 9.0.0b3.  It's possible that I'm getting my configuration wrong, but switching to Tcl/Tk 9.0.0b3 without this PR causes one failure, then adding these changes introduces a bunch of other failures, mostly around rounding as far as I can tell.

culler committed:
Add changes suggested by @zware

culler committed:
Remove trailing whitespace.

culler committed:
Deal with test_configure_width and test_configure_height for Notebooks.

culler committed:
Merge branch 'tk9_for_python314' of github.com:culler/cpython into tk9_for_python314

culler committed:
Need to make tests pass with 8.6.14 too, for the PR checks.

culler committed:
Fix version tests in NotebookTest

zware commented on Lib/test/test_tkinter/test_misc.py:
```diff
@@ -4,7 +4,7 @@
 import enum
 from test import support
-from test.test_tkinter.support import AbstractTkTest, AbstractDefaultRootTest, requires_tk
+from test.test_tkinter.support import AbstractTkTest, AbstractDefaultRootTest, requires_tk, tk_version
```
Nit: dangling unused import:

```suggestion
from test.test_tkinter.support import AbstractTkTest, AbstractDefaultRootTest, requires_tk
```

culler replied:
I pushed one more commit containing these changes and the ones above.

zware commented on Lib/test/test_tkinter/support.py:
```diff
@@ -112,6 +113,10 @@ def get_tk_patchlevel(root):
 
+pix_re = re.compile(r'[0-9]*\.?[0-9]*[cimp]{1}')
+def is_pixel_str(x):
+    return pix_re.fullmatch(x) != None
```
Nit: this doesn't seem to be used anymore.

zware reviewed (approved):
This is great, thank you!  I'm satisfied with this change overall, but I think we should get a once-over from @serhiy-storchaka for the tkinter changes and a final confirmation from @zooba that the `PCbuild/` changes match his expectations.

With your permission, I'd like to extract your ttk Notebook test fixes to a separate PR so that we can go ahead with it and get it backported.  (If you would prefer to make that split yourself, that works for me.)

I expect that we will eventually be backporting this whole PR even if we don't start shipping 9.0 with our <3.14 binaries because at some point our non-Windows CI systems will inevitably update to 9.0, but I'm not sure how quickly that will happen and want to be sure we get 8.6.15 supported as soon as we can.

culler committed:
Include changes suggested by @zware

zware committed:
Merge branch 'main' into tk9_for_python314

zooba commented on PCbuild/_tkinter.vcxproj:
```diff
@@ -112,6 +113,7 @@
     <_TclTkDLL Include="$(tcltkdir)\bin\$(tclDllName)" />
     <_TclTkDLL Include="$(tcltkdir)\bin\$(tkDllName)" />
     <_TclTkDLL Include="$(tcltkdir)\bin\$(tclZlibDllName)" />
+    <_TclTkDLL Include="$(tcltkdir)\bin\libtommath.dll" Condition="$(TclMajorVersion) == '9'"/>
```
Do we have to redistribute this file with Tcl/Tk 9.0? If so, there'll be updates under `Tools/msi` as well (look for references to tclZlibDllName).

Also, we should be using the name property here rather than the filename directly.

zooba replied:
Once we're using the name property, checking whether the name is empty rather than checking the version is going to be more robust.

```suggestion
    <_TclTkDLL Include="$(tcltkdir)\bin\$(tommathDLLName)" Condition="$(tommathDLLName) != ''"/>
```

culler replied:
Now using names, not paths and checking whether the name is empty in the Condition.

zooba commented on PCbuild/tcltk.props:
```diff
@@ -17,15 +17,20 @@
     <tclZlibDLLName>zlib1.dll</tclZlibDLLName>
-    <tcltkLib>$(tcltkDir)lib\tcl$(TclMajorVersion)$(TclMinorVersion)t$(TclDebugExt).lib;$(tcltkDir)lib\tk$(TkMajorVersion)$(TkMinorVersion)t$(TclDebugExt).lib</tcltkLib>
+    <tommathDLLName Condition="'$(TclMajorVersion)' == '9'">$(tcltkDir)bin\libtommath.dll</tommathDLLName>
+    <tommathLibName Condition="'$(TclMajorVersion)' == '9'">$(tcltkDir)lib\tommath.lib</tommathLibName>
```
These are names, not paths, so don't include the path.

```suggestion
    <tommathDLLName Condition="'$(TclMajorVersion)' == '9'">libtommath.dll</tommathDLLName>
    <tommathLibName Condition="'$(TclMajorVersion)' == '9'">tommath.lib</tommathLibName>
```

Add the path in the tcltkLib setting, or potentially just add `$(tcltkDir)Lib` to `_tkinter.vcxproj` as `AdditionalLibraryDirectories` and then the tcltkLib property can just contain filenames.

culler replied:
I ended up adding the path in tcltkLib.  It did not seem to work to add it to AdditionalLibraryDirectories, although I don't know why not.

zooba commented on PCbuild/tcltk.props:
```diff
@@ -17,15 +17,20 @@
     <tcltkDir Condition="$(tcltkDir) == ''">$(ExternalsDir)tcltk-$(TclVersion)\$(ArchName)\</tcltkDir>
     <tclWin32Exe Condition="$(Platform) == 'Win32'">$(tcltkDir)\bin\tclsh$(TclMajorVersion)$(TclMinorVersion)t.exe</tclWin32Exe>
     <tclWin32Exe Condition="$(Platform) != 'Win32'">$(tcltkDir)\..\win32\bin\tclsh$(TclMajorVersion)$(TclMinorVersion)t.exe</tclWin32Exe>
+    <tclExternalTommath Condition="$(TclMajorVersion) == '9'">TCL_WITH_EXTERNAL_TOMMATH;</tclExternalTommath>
```
This is not used anywhere, just remove it.

culler replied:
Actually, the TCL_WITH_EXTERNAL_TOMMATH constant **is** used when building against TclTk 9.0; although it is not used by 8.6.  With 9.0, if it is not defined then all of the symbols defined in libtommath.dll are reported by the linker as being undefined.  I believe that this is related to how the Tcl stubs library works;  it is not a straightforward linker error since libtommath.dll is being passed to the linker.

One might hope that the existence of a constant named TCL_WITH_EXTERNAL_TOMMATH indicates that one could choose not to have an external libtommath.dlll and instead arrange for libtommath to be statically linked into libtcl.dll, as is the case with Tcl 8.6.  However, I was not able to find a combination of build options which accomplished that.  I will try to find out if it is possible.

culler replied:
@zooba: As far as I can tell. the following command uses MSVC to build a static library tcl90s.lib which has both zlib and (part of) libtommath built-in:
 `nmake -f makefile.vc TCL_WITH_INTERNAL_ZLIB=1 TCL_WITH_EXTERNAL_TOMMATH=0 STATIC_BUILD=1`

According to dumpbin, the tclsh90s.exe executable built by that command does not have libtommath.dll or zlib.dll as dependencies.  So presumably the same would be true of the _tkinter extension if it were linked against tcl90s.lib. 

If you are willing to use gcc in msys64 then `configure --disable-static` seems to do the equivalent thing.

Reference: [https://core.tcl-lang.org/tips/doc/trunk/tip/538.md](https://core.tcl-lang.org/tips/doc/trunk/tip/538.md)

zooba reviewed (commented):
Along with the MSI changes (and possibly under PC/layout as well, I haven't checked), we need to make sure that we're okay with redistributing libtommath.dll and also that there isn't a static linking option for it. zlib unfortunately causes us trouble, but static linking here is impossible (because Tcl/Tk doesn't allow for it in their build scripts - I suspect the same will apply to tommath).

I'm okay with merging without the MSI changes, but we then cannot update any releases until it's done. As annoying as it is for backports, I'd require a comment in get_externals.bat to this effect.

culler committed:
Add changes to tcltk.props and _tkinter.vcxproj suggested by @zooba

culler committed:
Merge branch 'tk9_for_python314' of github.com:culler/cpython into tk9_for_python314

culler committed:
Merge branch 'main' into tk9_for_python314

zware committed:
Merge remote-tracking branch 'origin/main' into tk9_for_python314

zware committed:
Formatting fixups, remove note from build.bat

It's not really the place to call out TclVersion particularly.  We
probably should have a note about environment variables somewhere,
but I don't have a good wording for it at the moment.  We can add
that later on.
